fe1-web/src/features/wallet/objects/transaction/Transaction.ts

```diff
@@ -0,0 +1,387 @@
+
+    if (obj.transactionId === undefined) {
+      this.transactionId = this.hashTransaction();
+    } else {
+      if (obj.transactionId.valueOf() !== this.hashTransaction().valueOf()) {
+        throw new Error(
+          "The computed transaction hash does not correspond to the provided one when creating 'Transaction'",
+        );
+      }
+      this.transactionId = obj.transactionId;
+    }
+  }
+
+  /**
+   * Hashes the transaction to get its id
+   */
+  private readonly hashTransaction = (): Hash => {
+    // Recursively concatenating fields by lexicographic order of their names
+    const dataInputs = this.inputs.flatMap((input) => {
+      return [
+        input.script.publicKey.valueOf(),
+        input.script.signature.valueOf(),
+        input.script.type,
+        input.txOutHash.valueOf(),
+        input.txOutIndex.toString(),
+      ];
+    });
+    const dataOutputs = this.outputs.flatMap((output) => {
+      return [output.script.publicKeyHash.valueOf(), output.script.type, output.value.toString()];
+    });
+    const data = dataInputs
+      .concat([this.lockTime.toString()])
+      .concat(dataOutputs)
+      .concat([this.version.toString()]);
+
+    // Hash will take care of concatenating each fields length
+    return Hash.fromStringArray(...data);
+  };
+
+  /**
+   * Creates a transaction
+   * @param from the sender
+   * @param to the receiver
+   * @param currentBalance the current balance of the sender
+   * @param amount the amount to send to the receiver
+   * @param inputTransactions the transactions that contains the outputs to use as inputs in this transaction
+   */
+  public static create(
+    from: PopToken,
+    to: PublicKey,
+    currentBalance: number,
+    amount: number,
+    inputTransactions: TransactionState[],
+  ): Transaction {
+    const fromPublicKeyHash = Hash.fromPublicKey(from.publicKey);
+
+    const toPublicKeyHash = Hash.fromPublicKey(to);
+
+    const outputTo = {
+      value: amount,
+      script: {
+        type: STRINGS.script_type,
+        publicKeyHash: toPublicKeyHash.valueOf(),
+      },
+    };
+
+    const outputs: TransactionOutputState[] = [outputTo];
+
+    if (currentBalance > amount) {
+      // Send the rest of the value back to the owner, so that the entire balance
+      // is always in only one output
+      const outputFrom: TransactionOutputState = {
+        value: currentBalance - amount,
+        script: {
+          type: STRINGS.script_type,
+          publicKeyHash: fromPublicKeyHash.valueOf(),
+        },
+      };
+      outputs.push(outputFrom);
+    }
+
+    const inputs: Omit<TransactionInputState, 'script'>[] = Transaction.getInputsInToSign(
+      from.publicKey.valueOf(),
+      inputTransactions,
+    );
+
+    // Concatenate the data to sign
+    const dataString = Transaction.concatenateTxData(inputs, outputs);
+
+    // Sign with the popToken
+    const signature = from.privateKey.sign(Base64UrlData.encode(dataString));
```

Comment:
The general public key ? Are you talking about the `KeyPairStore.get()` in WalletHome ?

Reply:
No, what I mean is that you set `[k(COIN, POST_TRANSACTION), { signature: KEYPAIR }],` instead of `[k(COIN, POST_TRANSACTION), { signature: POP_TOKEN }],` in the message registry.

Reply:
(Which is fine since here you add a pop token signature, was just confused at first why you are doing it this way and I then assumed it would be because `[k(COIN, POST_TRANSACTION), { signature: POP_TOKEN }]` will not allow you to use a specific pop token)

Reply:
Oh okay I missed the 'the message' part, in fact I made it this way, because at first the coinbase transaction messages would have needed a signature from the organizer, but it's not the case anymore (the organizer's signature is still needed but inside the transaction itself), so it does not really matter now